packages/pyright-internal/src/common/uri/baseUri.ts

```diff
@@ -0,0 +1,288 @@
+    get extname(): string {
+        return this.getExtnameImpl();
+    }
+
+    // Returns a URI where the path just contains the root folder.
+    get root(): Uri {
+        return this.getRootImpl();
+    }
+
+    // Returns a URI where the path contains the path with .py appended.
+    get packageUri(): Uri {
+        // This is assuming that the current path is a directory already.
+        return this.addPath('.py');
+    }
+
+    // Returns a URI where the path contains the path with .pyi appended.
+    get packageStubUri(): Uri {
+        // This is assuming that the current path is a directory already.
+        return this.addPath('.pyi');
+    }
+
+    // Returns a URI where the path has __init__.py appended.
+    get initPyUri(): Uri {
+        // This is assuming that the current path is a directory already.
+        return this.combinePaths('__init__.py');
+    }
+
+    // Returns a URI where the path has __init__.pyi appended.
+    get initPyiUri(): Uri {
+        // This is assuming that the current path is a directory already.
+        return this.combinePaths('__init__.pyi');
+    }
+
+    // Returns a URI where the path has py.typed appended.
+    get pytypedUri(): Uri {
+        // This is assuming that the current path is a directory already.
+        return this.combinePaths('py.typed');
+    }
+
+    isEmpty(): boolean {
+        return false;
+    }
+
+    abstract toString(): string;
+
+    abstract toUserVisibleString(): string;
+
+    abstract matchesRegex(regex: RegExp): boolean;
+
+    replaceExtension(ext: string): Uri {
+        const dir = this.getDirectory();
+        const base = this.filename;
+        const newBase = base.slice(0, base.length - this.extname.length) + ext;
+        return dir.combinePaths(newBase);
+    }
+
+    addExtension(ext: string): Uri {
+        return this.addPath(ext);
+    }
+
+    abstract addPath(extra: string): Uri;
+
+    // Returns a URI where the path is the directory name of the original URI, similar to the UNIX dirname command.
+    getDirectory(): Uri {
+        return this.getDirectoryImpl();
+    }
+
+    getRootPathLength(): number {
+        return this.getRootPath().length;
+    }
+
+    // Determines whether a path consists only of a path root.
+    abstract isRoot(): boolean;
+
+    // Determines whether a Uri is a child of some parent Uri.
+    abstract isChild(parent: Uri, ignoreCase?: boolean): boolean;
+
+    abstract isLocal(): boolean;
+
+    isUntitled(): boolean {
+        return this.scheme === 'untitled';
+    }
+
+    equals(other: Uri | undefined, ignoreCase?: boolean): boolean {
+        if (ignoreCase) {
+            return this.key.toLowerCase() === other?.key.toLowerCase();
+        }
+        return this.key === other?.key;
+    }
+
+    abstract startsWith(other: Uri | undefined, ignoreCase?: boolean): boolean;
+
+    pathStartsWith(name: string): boolean {
+        // ignore path separators.
+        name = normalizeSlashes(name);
```

Comment:
now, with everything changed to `Uri`, are there any case where we need to explicitly use `normalizeSlash` outside of `file system implementation`? 

in Uri, `/` is always used regardless of underlying file system?

Reply:
Uri does not use `/` everywhere. Just the WebURI does. FileURI still uses whatever slash is appropriate for the file system.